### PR TITLE
Generate object ids for plasma error handling

### PR DESCRIPTION
**Pull request by hkvision**
Previously if putting into plasma tasks fail, each spark retrial would create a new object id to put the data. This would lead to memory leak if only some tasks fail.

Generate an ObjectID for each partition beforehand for plasma.
In case tasks fail, spark would retry to put into plasma. If the target id already exists, we suppose that the task on that partition succeeds and does nothing. Otherwise, still use this id to put into plasma.

### CODE REVIEW

**Comment by hkvision** [comment]
What about this? @jason-dai @yangw1234

**Comment by hkvision** [comment]
jenkins passed: http://10.239.47.210:18888/job/ZOO-PR-Validation/3683/

File: pyzoo/zoo/orca/data/shard.py [hunk 1]
```diff
@@ -322,33 +322,28 @@ def get_data(data):
...
 
-        # Generate a random string here to make sure that when this method is called twice, the
-        # seeds to generate plasma ObjectID are different.
-        random_str = ''.join(
-            [random.choice(string.ascii_letters + string.digits) for i in range(32)])
-        object_id_node_ips = self.rdd.mapPartitionsWithIndex(put_to_plasma(random_str)).collect()
+        # Create plasma ObjectIDs beforehand instead of creating a random one every time.
+        import pyarrow.plasma as plasma
+        object_ids = [plasma.ObjectID.from_random() for i in range(self.rdd.getNumPartitions())]
+        object_id_node_ips = self.rdd.mapPartitionsWithIndex(put_to_plasma(object_ids)).collect()
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/shard.py:346]: Is it possible to generate ID that already exists in the executor side?

Comment (hkvision) [reply pyzoo/zoo/orca/data/shard.py:346]: Can you give more details? Originally on each node, there would be no IDs. If some tasks succeed, there will be IDs and can be retrieved.

Comment (jason-dai) [reply pyzoo/zoo/orca/data/shard.py:346]: How do you make sure the generated ID does not already exist in the plasma?

Comment (hkvision) [reply pyzoo/zoo/orca/data/shard.py:346]: Object ID is a byte string of length 20 that consists of numbers and lower-letter chars. So I suppose there is little probability to generate two same random IDs?

Comment (jason-dai) [reply pyzoo/zoo/orca/data/shard.py:346]: How does each previously plasma id get generated in each partition?

Comment (hkvision) [reply pyzoo/zoo/orca/data/shard.py:346]: Previously if we don't specify the ID, plasma itself would generate a random ID for the data. Now if we specify the ID, plasma would exactly use the id we input.

Comment (jason-dai) [reply pyzoo/zoo/orca/data/shard.py:346]: > Previously if we don't specify the ID, plasma itself would generate a random ID for the data. Now if we specify the ID, plasma would exactly use the id we input.

How does plasma generate the ID?

Comment (hkvision) [reply pyzoo/zoo/orca/data/shard.py:352]: I suppose plasma would also randomly generate the ID of the expected format.

**Comment by hkvision** [comment]
Synced with @jason-dai offline. Would merge it first. If there are further issue, would fix in another PR.

**Comment by hkvision** [comment]
jenkins http://10.239.47.210:18888/job/ZOO-PR-Validation/3705/
One failure in scala is irrelevant.

**Comment by hkvision** [comment]
jenkins passed: http://10.239.47.210:18888/job/ZOO-PR-Validation/3707/

**Pull request merged**